CppCoreGuidelines.md

```diff
@@ -19254,6 +19255,41 @@ An API class and its members can't live in an unnamed namespace; but any "helper
 
 * ???
 
+### <a name="Rs-forward"></a>SF.23: Keep forward declarations independent
+
+##### Reason
+
+Reading a forward declaration is easier since all the needed information is on the one line
+Refactoring is easier because when a class moves namespaces only the lines that forward declare it must be changed. Renaming a namespace is still just a find and replace.
+Merging is easier because each line is independent
+Writing is easier because it's easier to decide where to insert (because it doesn't make a difference)
+
+##### Example
+
+Bad
+
+    namespace MyNamespace {
+    namespace InnerNamespace {
+    class C1;
+    class C2;
+    }
+    class C3;
+    }
+
+Good
+
+    namespace MyNamespace::InnerNamespace { class C1; }
+    namespace MyNamespace::InnerNamespace { class C2; }
+    namespace MyNamespace { class C3; }
+
+##### Note
+
+This assumes that namespace names are sufficiently short and nesting is sufficiently shallow to avoid violating any line length limits. If this is not the case, it may indicate namespace names are too long or deeply nested.
```

Comment:
s/deeply/too deeply/
Is there a guideline "Don't nest namespaces too deeply" that could be cross-referenced here?
This also assumes C++17, because `namespace X::Y` wasn't legal before C++17. Do we care? (Probably not.)

Reply:
I'm not aware of a "Don't nest namespaces too deeply" rule in the core guidelines. I think it's a good rule but it's subjective and potentially controversial. I do think that if you can't fit a fully qualified class name (templates aside) on one line then in all likelihood your namespaces are too long/nested. I'm careful to use "may" here though because it's going to depend. Basically if someone says "I can't follow this rule and adhere to my line length rules" I want to offer some sort of solution.

That's a really good point about c++17. This guideline is completely inapplicable pre c++17. I don't think that makes it bad but I think it's definitely worth noting in the text as it could save someone using c++14 a lot of time.